operators/src/upgrade/phases/init.rs

```diff
@@ -0,0 +1,247 @@
+use std::collections::HashMap;
+
+use crate::upgrade::k8s::crd::v0::UpgradePhase;
+
+/// Core components state.
+pub struct ComponentsState {
+    core: CoreComponentsState,
+    supportability: SupportabilityComponentsState,
+    tracing: TracingComponentsState,
+}
+
+impl ComponentsState {
+    /// Core components state.
+    pub(crate) fn core(&self) -> CoreComponentsState {
+        self.core.clone()
+    }
+
+    /// Supportability components state.
+    pub(crate) fn supportability(&self) -> SupportabilityComponentsState {
+        self.supportability.clone()
+    }
+
+    /// tracing components state.
+    pub(crate) fn tracing(&self) -> TracingComponentsState {
+        self.tracing.clone()
+    }
+
+    /// Initialize state.
+    pub(crate) fn with_state(upgrade_phase: UpgradePhase) -> Self {
+        Self {
+            core: CoreComponentsState::with_state(upgrade_phase.clone()),
+            supportability: SupportabilityComponentsState::with_state(upgrade_phase.clone()),
+            tracing: TracingComponentsState::with_state(upgrade_phase),
+        }
+    }
+
+    /// Covert into hashmaps.
+    pub(crate) fn convert_into_hash(&self) -> HashMap<String, HashMap<String, UpgradePhase>> {
+        let mut h: HashMap<String, HashMap<String, UpgradePhase>> = HashMap::new();
+        h.insert("core".to_string(), self.core().convert_into_hash());
+        h.insert(
+            "supportability".to_string(),
+            self.supportability().convert_into_hash(),
+        );
+        h.insert("tracing".to_string(), self.tracing().convert_into_hash());
+        h
+    }
+}
+
+/// Core components state.
+#[derive(Clone)]
+pub struct CoreComponentsState {
+    core_agent: UpgradePhase,
+    rest_api: UpgradePhase,
+    csi_controller: UpgradePhase,
+    csi_node: UpgradePhase,
+    io_engine: UpgradePhase,
+    metrics_exporter_pool: UpgradePhase,
+    disk_pool_operator: UpgradePhase,
```

Comment:
missing the ha_node?
Also I forgot what was the plan with the persistent store (etcd). In which phase do we update it if we need to, at each component level?

Reply:
There's no component level logic in this operators, all components get the same state (success is all succeeded, and failure otherwise), irrespective of what happened to each one of them. 

For this operator, etc or other dependent charts update during the UpdatingControlPlane State. Helm will only update it if the dependent chart version is different from the one that is installed.

Reply:
So where will the logic parts go?

Reply:
This is a part of the live and hot upgrade operator, and not this one.

Reply:
Ah I see, thanks for explaining :+1: